features/opennms-es-rest/src/main/java/org/opennms/plugins/elasticsearch/rest/EventToIndex.java

```diff
@@ -384,9 +384,14 @@ private Index createEventIndexFromEvent(final Event event, final IndexAndType in
 
 		body.put("host",event.getHost());
 
-		//get params from event
 		for(Parm parm : event.getParmCollection()) {
-			body.put("p_" + parm.getParmName(), parm.getValue().getContent());
+			// We have to handle oid parameters differently, as elastic would consider a containing "." as a sub
+			// field, e.g. ".1.2" results in a mapping of { 1: properties: { 2 } } which may exceed the maximum allowed
+			// fields for a mapping, see NMS-9831. Initially an array was used, to store the oids in the fashion:
+			// { "oid": "1.1.1", "value": "dummy value" } but elastic has some limitations on accessing objects in an
```

Comment:
Could you explain the limitations you are referring to?

Replacing periods with underscores will address the default depth issue of a parameter within elastic. As flat parameters, the 1000 default limit will still be hit, the array addresses both the depth and parameter limit.

A corresponding updated mapping template would allow you to "keyword" the "oid" and "value" values in addition can also be used to create a corresponding ".raw" sub-field that is type "text" for full text searches.

Reply:
> Could you explain the limitations you are referring to?

This is what the official elastic documentation says about arrays: https://www.elastic.co/guide/en/elasticsearch/reference/current/array.html

> Replacing periods with underscores will address the default depth issue of a parameter within elastic. As flat parameters, the 1000 default limit will still be hit, the array addresses both the depth and parameter limit. [...]

That is true and not true. Replacing periods with underscores will flatten the parameters as you said and will reduce the likelihood of hitting the limit. As the event parameters can be configured and therefore are also dynamic only putting all event parameters in an array would fully fix the issue. 

I opted in for a consistent implementation as replacing `.` with `_` should probably always be performed and not only for `oid` parameters. Putting all parameters in a parameter array is not possible due to the limitations mentioned above (HZN-1272 is really the problem here). We can look into the `nested` option for arrays, though. 

Just a side note: The limit is configurable: https://discuss.elastic.co/t/total-fields-limit-setting/53004

Reply:
How about a middle of the road solution, I would suggest adding a new parameter for the plugin to be able to chose between nested or flat.

In my instance, I would have to deal with over 20K+ flat fields per indices which can influence performance and memory consumption within elasticsearch. With a option available for choosing nested, I would eliminate this field explosion in my situation, while still preserving the option for others not in a similar situation to choose flat.

Reply:
Yeah, I see your point now.